File: 백준/Gold/20955. 민서의 응급 수술/민서의 응급 수술.py
"""
트리로 연결이면 연결요소의 개수 구하기

연산횟수 = 연결요소 개수 - 1

그런데 사이클이 있다면?

연산횟수 = 연결요소 개수 - 1 + 사이클 개수
"""
import sys
from collections import deque

input = lambda: sys.stdin.readline().strip()

N, M = map(int, input().split())

edge = [[] for _ in range(N + 1)]
visited = [False for _ in range(N + 1)]

for _ in range(M):
    u, v = map(int, input().split())
    edge[u].append(v)
    edge[v].append(u)


def bfs(start):
    cnt = 0
    q = deque()
    visited[start] = True
    q.append((start, -1))

    while q:
        cur, prev = q.popleft()
        for nxt in edge[cur]:
            if nxt == prev:
                continue
            if not visited[nxt]:
                visited[nxt] = True
                q.append((nxt, cur))
            else:
                cnt += 1
    return cnt // 2


cnt_component = 0
cnt_cycle = 0
for i in range(1, N + 1):
    if not visited[i]:
        cnt_component += 1
        cnt_cycle += bfs(i)

print(cnt_component - 1 + cnt_cycle)

File: 백준/Gold/20955. 민서의 응급 수술/민서의 응급 수술.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    private static int N, M;
    private static List<Integer>[] edge;
    private static boolean[] visited;

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        edge = new ArrayList[N + 1];

        for (int i = 1; i < N + 1; i++) {
            edge[i] = new ArrayList<Integer>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            edge[u].add(v);
            edge[v].add(u);
        }

        visited = new boolean[N + 1];
        int cnt_component = 0;
        int cnt_cycle = 0;
        for (int i = 1; i < N + 1; i++) {
            if (!visited[i]) {
                cnt_cycle += bfs(i);
                cnt_component += 1;
            }
        } // end of for i

        System.out.println(cnt_component + cnt_cycle - 1);
        
    } // end of main

    private static int bfs(int start) {
        int cnt = 0;
        Queue<Integer> q = new ArrayDeque<>();

        q.offer(start);
        q.offer(-1);
        visited[start] = true;

        while (!q.isEmpty()) {
            int cur = q.poll();
            int prev = q.poll();
            for (int nxt: edge[cur]) {
                if (nxt == prev) continue;
                if (!visited[nxt]) {
                    q.offer(nxt);
                    q.offer(cur);
                    visited[nxt] = true;
                } else {
                    cnt += 1;
                }
            }
        }

        return cnt / 2;
    } // end of bfs
} // end of class